include/arch/arm/arch/machine/gic_v3.h

```diff
@@ -299,4 +333,209 @@ void ipi_send_target(irq_t irq, word_t cpuTargetList);
+        SYSTEM_READ_WORD(ICH_LR0(5), lr);
+        break;
+    case 6:
+        SYSTEM_READ_WORD(ICH_LR0(6), lr);
+        break;
+    case 7:
+        SYSTEM_READ_WORD(ICH_LR0(7), lr);
+        break;
+    case 8:
+        SYSTEM_READ_WORD(ICH_LR8(0), lr);
+        break;
+    case 9:
+        SYSTEM_READ_WORD(ICH_LR8(1), lr);
+        break;
+    case 10:
+        SYSTEM_READ_WORD(ICH_LR8(2), lr);
+        break;
+    case 11:
+        SYSTEM_READ_WORD(ICH_LR8(3), lr);
+        break;
+    case 12:
+        SYSTEM_READ_WORD(ICH_LR8(4), lr);
+        break;
+    case 13:
+        SYSTEM_READ_WORD(ICH_LR8(5), lr);
+        break;
+    case 14:
+        SYSTEM_READ_WORD(ICH_LR8(6), lr);
+        break;
+    case 15:
+        SYSTEM_READ_WORD(ICH_LR8(7), lr);
+        break;
+    default:
+        fail("GICv3: Invalid ICH_LR_EL2");
+        break;
+    }
+    virq.words[0] = lr;
+    return virq;
+}
+
+static inline void set_gic_vcpu_ctrl_lr(word_t n, virq_t lr)
+{
+    switch (n) {
+    case 0:
+        SYSTEM_WRITE_WORD(ICH_LR0(0), lr.words[0]);
+        break;
+    case 1:
+        SYSTEM_WRITE_WORD(ICH_LR0(1), lr.words[0]);
+        break;
+    case 2:
+        SYSTEM_WRITE_WORD(ICH_LR0(2), lr.words[0]);
+        break;
+    case 3:
+        SYSTEM_WRITE_WORD(ICH_LR0(3), lr.words[0]);
+        break;
+    case 4:
+        SYSTEM_WRITE_WORD(ICH_LR0(4), lr.words[0]);
+        break;
+    case 5:
+        SYSTEM_WRITE_WORD(ICH_LR0(5), lr.words[0]);
+        break;
+    case 6:
+        SYSTEM_WRITE_WORD(ICH_LR0(6), lr.words[0]);
+        break;
+    case 7:
+        SYSTEM_WRITE_WORD(ICH_LR0(7), lr.words[0]);
+        break;
+    case 8:
+        SYSTEM_WRITE_WORD(ICH_LR8(0), lr.words[0]);
+        break;
+    case 9:
+        SYSTEM_WRITE_WORD(ICH_LR8(1), lr.words[0]);
+        break;
+    case 10:
+        SYSTEM_WRITE_WORD(ICH_LR8(2), lr.words[0]);
+        break;
+    case 11:
+        SYSTEM_WRITE_WORD(ICH_LR8(3), lr.words[0]);
+        break;
+    case 12:
+        SYSTEM_WRITE_WORD(ICH_LR8(4), lr.words[0]);
+        break;
+    case 13:
+        SYSTEM_WRITE_WORD(ICH_LR8(5), lr.words[0]);
+        break;
+    case 14:
+        SYSTEM_WRITE_WORD(ICH_LR8(6), lr.words[0]);
+        break;
+    case 15:
+        SYSTEM_WRITE_WORD(ICH_LR8(7), lr.words[0]);
+        break;
+    default:
+        fail("GICv3: Invalid ICH_LR_EL2");
+        break;
+
+    }
+}
+
+#define GIC_INVALID_IRQ_IDX 0
+static inline int get_vgic_irq_idx(void) {
+    int irq_idx = GIC_INVALID_IRQ_IDX;
+    uint32_t eisr0 = get_gic_vcpu_ctrl_eisr0();
+
+    for (int i = 0; i < gic_vcpu_num_list_regs; i++) {
+        if (eisr0 & BIT(i)) {
+            irq_idx |= BIT(i);
+        }
+    }
+    return irq_idx;
+}
+
+static inline virq_t gic_virq_pending_new(word_t group, word_t priority, word_t vid)
+{
+    return virq_virq_pending_new(0, group, priority, 0x200, vid);
+}
+
+static inline bool_t gic_valid_irq_idx(int irq_idx)
+{
+    return irq_idx < gic_vcpu_num_list_regs;
```

Comment:
given that irq_idx could be a result of | several bits, it could be larger than the gic_vcpu_num_list_regs?